File: aurora_genesis_core/fractal/sierpinski.py
'''Sierpinski simplex centroid generator. GPU-accelerated deterministic fractal initialization.'''

from __future__ import annotations

import math
from typing import Optional

try:
    import torch
    _TORCH_AVAILABLE = True
except ImportError:
    _TORCH_AVAILABLE = False
    torch = None  # type: ignore


if not _TORCH_AVAILABLE:
    raise ImportError(
        "aurora_genesis_core.fractal.sierpinski requires torch. "
        "Install project dependencies so `torch` is available (see docs/handover.md)."
    )


def regular_simplex_vertices(n_dim: int, device: Optional[str] = None) -> torch.Tensor:
    """
    Generate vertices of a regular N-simplex centered at origin.

    Algorithm: Place first vertex at (1, 0, ..., 0), then use Gram-Schmidt
    to generate orthogonal unit vectors. The simplex vertices are then
    normalized to lie on the unit sphere.

    Args:
        n_dim: Number of dimensions (N-simplex has N+1 vertices).
        device: PyTorch device ('cuda', 'cpu', etc.). Defaults to 'cuda' if available.

    Returns:
        Tensor of shape (n_dim + 1, n_dim) with vertices on unit sphere.

    Raises:
        ValueError: If n_dim < 1.
    """
    if n_dim < 1:
        raise ValueError(f"n_dim must be >= 1, got {n_dim}")

    if device is None:
        device = "cuda" if torch.cuda.is_available() else "cpu"

    # For small dimensions, use explicit formulas
    if n_dim == 1:
        # 1D simplex (line segment): vertices at -1 and +1
        vertices = torch.tensor([[-1.0], [1.0]], device=device)
    elif n_dim == 2:
        # 2D simplex (triangle): equilateral triangle
        sqrt3 = math.sqrt(3.0)
        vertices = torch.tensor(
            [[-1.0, -sqrt3 / 3], [1.0, -sqrt3 / 3], [0.0, 2 * sqrt3 / 3]],
            device=device,
        )
        # Normalize to unit circle
        norms = torch.norm(vertices, dim=1, keepdim=True)
        vertices = vertices / (norms + 1e-8)
    elif n_dim == 3:
        # 3D simplex (tetrahedron): regular tetrahedron
        sqrt3 = math.sqrt(3.0)
        sqrt6 = math.sqrt(6.0)
        vertices = torch.tensor(
            [
                [-1.0, -1.0 / sqrt3, -1.0 / sqrt6],
                [1.0, -1.0 / sqrt3, -1.0 / sqrt6],
                [0.0, 2.0 / sqrt3, -1.0 / sqrt6],
                [0.0, 0.0, 3.0 / sqrt6],
            ],
            device=device,
        )
        # Normalize to unit sphere
        norms = torch.norm(vertices, dim=1, keepdim=True)
        vertices = vertices / (norms + 1e-8)
    else:
        # General case: use QR decomposition to generate orthogonal vectors
        # Start with identity-like matrix, then orthogonalize
        eye_like = torch.eye(n_dim + 1, n_dim, device=device)
        # First vertex at (1, 0, ..., 0)
        eye_like[0, 0] = 1.0
        # Subsequent vertices shifted to form simplex
        for i in range(1, n_dim + 1):
            eye_like[i, i - 1] = 1.0 - (1.0 / (n_dim + 1))
            eye_like[i, :i] = -1.0 / (n_dim + 1)

        # Normalize to unit sphere
        norms = torch.norm(eye_like, dim=1, keepdim=True)
        vertices = eye_like / (norms + 1e-8)

    return vertices


def generate_sierpinski_centroids(
    n_dim: int,
    depth: int,
    n_centroids: Optional[int] = None,
    seed: int = 1234,
    device: Optional[str] = None,
) -> torch.Tensor:
    """
    Generate Sierpinski N-simplex centroids using chaos game iteration.

    Algorithm:
    1. Generate regular N-simplex vertices using `regular_simplex_vertices()`.
    2. For each centroid (determined by `depth`), perform chaos game:
       - Start at origin.
       - For `depth` iterations:
         - Choose a random vertex (via seeded LCG).
         - Move to midpoint between current point and chosen vertex.
    3. Returns deterministic set of centroids on GPU.

    LCG: state = (state * 1664525 + 1013904223) % 2^32
    (Matches CUDA header `lib/fractal_sierpinski.h`)

    Args:
        n_dim: Number of dimensions (determines N-simplex type).
        depth: Number of chaos game iterations per centroid.
        seed: Random seed for LCG (ensures determinism).
        device: PyTorch device ('cuda', 'cpu', etc.). Defaults to 'cuda' if available.

    Returns:
        Tensor of shape (2^(n_dim * depth), n_dim) with centroids on GPU.
        NOTE: For large n_dim * depth, this is intractable. Use depth=1-2 and downsample.

    Raises:
        ValueError: If n_dim < 1 or depth < 1.
    """
    if n_dim < 1:
        raise ValueError(f"n_dim must be >= 1, got {n_dim}")
    if depth < 1:
        raise ValueError(f"depth must be >= 1, got {depth}")

    if device is None:
        device = "cuda" if torch.cuda.is_available() else "cpu"

    # Generate simplex vertices
    vertices = regular_simplex_vertices(n_dim, device=device)
    n_vertices = vertices.shape[0]  # n_dim + 1

    # Number of centroids: 2^(n_dim * depth) is intractable for large values.
    # For practical use, allow callers (e.g. memory init) to request exactly N.
    # If N is not provided, we fall back to the original behavior with a cap.
    max_centroids = 65536
    if n_centroids is None:
        n_centroids_target = 2 ** (n_dim * depth)
        if n_centroids_target > max_centroids:
            # Use depth=1 and generate max_centroids instead (bounded runtime).
            depth_actual = 1
            n_centroids_effective = max_centroids
        else:
            depth_actual = depth
            n_centroids_effective = n_centroids_target
    else:
        if n_centroids < 1:
            raise ValueError(f"n_centroids must be >= 1, got {n_centroids}")
        depth_actual = depth
        n_centroids_effective = min(int(n_centroids), max_centroids)

    # LCG constants (matches CUDA header)
    LCG_MULT = 1664525
    LCG_ADD = 1013904223

    # Vectorized chaos game:
    # Maintain one RNG state per centroid and update all points in parallel.
    idx = torch.arange(n_centroids_effective, device=device, dtype=torch.int64)
    state = ((seed & 0xFFFFFFFF) ^ ((idx * 747796405) & 0xFFFFFFFF)) & 0xFFFFFFFF

    centroids = torch.zeros(
        (n_centroids_effective, n_dim),
        device=device,
        dtype=torch.float32,
    )

    for _ in range(depth_actual):
        state = ((state * LCG_MULT) + LCG_ADD) & 0xFFFFFFFF
        vertex_idx = torch.remainder(state, n_vertices).to(dtype=torch.long)
        centroids = (centroids + vertices[vertex_idx]) * 0.5

    return centroids

File: src/aurora/fractal/sierpinski.ts
/**
 * Sierpinski N‑simplex fractal centroid generator (reference implementation).
 * Based on Aurora‑Genesis fractal/sierpinski.py (chaos‑game iteration).
 */

export interface SierpinskiOptions {
  /** Number of dimensions (N‑simplex has N+1 vertices) */
  nDim: number;
  /** Depth of chaos‑game iteration */
  depth: number;
  /** Random seed for deterministic generation */
  seed?: number;
  /** Maximum number of centroids to generate (if depth would produce too many) */
  maxCentroids?: number;
}

/**
 * Generate vertices of a regular N‑simplex centered at origin.
 */
function regularSimplexVertices(nDim: number): number[][] {
  if (nDim < 1) throw new Error(`nDim must be >= 1, got ${nDim}`);

  // Small dimensions: explicit formulas
  if (nDim === 1) {
    return [[-1.0], [1.0]];
  }
  if (nDim === 2) {
    const sqrt3 = Math.sqrt(3);
    return [
      [-1.0, -sqrt3 / 3],
      [1.0, -sqrt3 / 3],
      [0.0, 2 * sqrt3 / 3],
    ];
  }
  if (nDim === 3) {
    const sqrt3 = Math.sqrt(3);
    const sqrt6 = Math.sqrt(6);
    return [
      [-1.0, -1.0 / sqrt3, -1.0 / sqrt6],
      [1.0, -1.0 / sqrt3, -1.0 / sqrt6],
      [0.0, 2.0 / sqrt3, -1.0 / sqrt6],
      [0.0, 0.0, 3.0 / sqrt6],
    ];
  }

  // General case: generate using QR decomposition analogue
  const vertices: number[][] = [];
  for (let i = 0; i < nDim + 1; i++) {
    const vertex: number[] = new Array(nDim).fill(-1.0 / (nDim + 1));
    if (i === 0) {
      vertex[0] = 1.0;
    } else {
      vertex[i - 1] = 1.0 - (1.0 / (nDim + 1));
    }
    vertices.push(vertex);
  }

  // Normalize to unit sphere
  for (let i = 0; i < vertices.length; i++) {
    const vertex = vertices[i];
    const norm = Math.sqrt(vertex.reduce((sum, v) => sum + v * v, 0));
    if (norm > 0) {
      for (let j = 0; j < nDim; j++) {
        vertex[j] /= norm;
      }
    }
  }

  return vertices;
}

/**
 * Generate Sierpinski N‑simplex centroids using chaos‑game iteration.
 * Deterministic, based on LCG random sequence.
 */
export function generateSierpinskiCentroids(
  options: SierpinskiOptions
): number[][] {
  const { nDim, depth, seed = 1234, maxCentroids = 65536 } = options;

  if (nDim < 1) throw new Error(`nDim must be >= 1, got ${nDim}`);
  if (depth < 1) throw new Error(`depth must be >= 1, got ${depth}`);

  // Generate simplex vertices
  const vertices = regularSimplexVertices(nDim);
  const nVertices = vertices.length; // nDim + 1

  // Number of centroids: 2^(nDim * depth) is intractable for large values
  let nCentroidsTarget = 2 ** (nDim * depth);
  if (nCentroidsTarget > maxCentroids) {
    // Use depth=1 and generate maxCentroids instead (bounded runtime)
    nCentroidsTarget = maxCentroids;
  }

  const nCentroids = Math.min(nCentroidsTarget, maxCentroids);

  // LCG constants (matches Aurora CUDA header)
  const LCG_MULT = 1664525;
  const LCG_ADD = 1013904223;

  // Initialize centroids at origin
  const centroids: number[][] = Array.from({ length: nCentroids }, () => 
    new Array(nDim).fill(0)
  );

  // Initialize LCG states per centroid
  const states: number[] = Array.from({ length: nCentroids }, (_, idx) => {
    return ((seed & 0xFFFFFFFF) ^ ((idx * 747796405) & 0xFFFFFFFF)) & 0xFFFFFFFF;
  });

  // Chaos‑game iteration
  for (let iter = 0; iter < depth; iter++) {
    for (let c = 0; c < nCentroids; c++) {
      // Update LCG state
      states[c] = (((states[c] >>> 0) * LCG_MULT) + LCG_ADD) & 0xFFFFFFFF;
      // Choose random vertex
      const vertexIdx = (states[c] >>> 0) % nVertices;
      if (vertexIdx < 0 || vertexIdx >= vertices.length) throw new Error(`vertexIdx out of bounds: ${vertexIdx}, vertices length ${vertices.length}`);
      const vertex = vertices[vertexIdx];
      if (!vertex) throw new Error(`vertex is undefined at index ${vertexIdx}`);

      // Move to midpoint between current centroid and chosen vertex
      const centroid = centroids[c];
      for (let d = 0; d < nDim; d++) {
        centroid[d] = (centroid[d] + vertex[d]) * 0.5;
      }
    }
  }

  return centroids;
}

/**
 * Generate centroids and map them to a specific range (e.g., [-1, 1]^N).
 */
export function generateSierpinskiCentroidsNormalized(
  options: SierpinskiOptions,
  rangeMin: number = -1,
  rangeMax: number = 1
): number[][] {
  const centroids = generateSierpinskiCentroids(options);

  // Find bounding box
  let minVal = Infinity;
  let maxVal = -Infinity;
  for (const centroid of centroids) {
    for (const val of centroid) {
      if (val < minVal) minVal = val;
      if (val > maxVal) maxVal = val;
    }
  }

  // Normalize to range
  const scale = (rangeMax - rangeMin) / (maxVal - minVal);
  const offset = rangeMin - minVal * scale;

  return centroids.map(centroid => 
    centroid.map(val => val * scale + offset)
  );
}

/**
 * Utility: compute Euclidean distance between two vectors.
 */
export function euclideanDistance(a: number[], b: number[]): number {
  if (a.length !== b.length) throw new Error(`Vector length mismatch: ${a.length} vs ${b.length}`);
  let sum = 0;
  for (let i = 0; i < a.length; i++) {
    const diff = a[i] - b[i];
    sum += diff * diff;
  }
  return Math.sqrt(sum);
}

/**
 * Find nearest centroid for a given vector.
 */
export function findNearestCentroid(
  vector: number[],
  centroids: number[][]
): { index: number; distance: number } {
  if (centroids.length === 0) throw new Error('centroids array is empty');
  if (vector.length !== centroids[0].length) throw new Error('vector dimension mismatch');
  let bestIdx = 0;
  let bestDist = Infinity;

  for (let i = 0; i < centroids.length; i++) {
    if (centroids[i].length !== vector.length) throw new Error(`centroid ${i} dimension mismatch`);
    const dist = euclideanDistance(vector, centroids[i]);
    if (dist < bestDist) {
      bestDist = dist;
      bestIdx = i;
    }
  }

  return { index: bestIdx, distance: bestDist };
}

// ADID_ROLLBACK (from adm.exe)
// SDID_ROLLBACK {
//   "target_file": "D:\\zPython\\grok-cli\\src/aurora/fractal/sierpinski.ts"
//   "update_script": "adm.exe"
//   "backup_path": "none"
//   "created_at": "2026-03-01T14:37:42.424757+00:00"
//   "new_hash": "7425f8797db330ac370faed34e266dcd"
//   "goal_id": "create_src_aurora_fractal_sierpinski_ts"
//   "semantics": "Create src/aurora/fractal/sierpinski.ts (Aurora reference implementation)"
//   "update_attrs": {"relative_path": "src/aurora/fractal/sierpinski.ts", "update_type": "text", "mode": "overwrite", "encoding": "utf-8", "find_pattern": null, "find_text": "", "replace_present": true}
//   "restore_cmd": "uv run adm --rollback \"D:\\zPython\\grok-cli\\src/aurora/fractal/sierpinski.ts\""
// }
